# Automatização de mensagens

Nesse notebook, vamos realizar a automatização de mensagens no **Gmail**, através das bibliotecas [*PyAutoGui*](https://pyautogui.readthedocs.io/en/latest/) e [*PyPerClip*](https://pypi.org/project/pyperclip/), no intuito de visitar um endereço de uma pasta compartilhada do drive, realizar o download de uma planilha, realizar o cálculo e enviar para um email.

Primeiro, vamos realizar comandos básicos de atalhos do teclado para o navegador, no intuito de ensinar cada comando da biblioteca. Assim, vamos acessar, de maneira automatizada, o Gmail.

A seguir, vamos clicar no botão "Escrever", para inserir os campos de *email, assunto e corpo do email.*

A documentação para as bibliotecas pode ser acessar através de [PyAutoGui](https://pyautogui.readthedocs.io/en/latest/) e [PyPerClip](https://pypi.org/project/pyperclip/). Os principais comandos que serão utilizados serão:

* **size**, da biblioteca PyAutoGui, para exibir o tamanho da tela;
* **hotkey**, da biblioteca PyAutoGui, para simular comandos do teclado;
* **write**, da biblioteca PyAutoGui, para escrever;
* **press**, da biblioteca PyAutoGui, para simular pressionar uma tecla;
* **position**, da biblioteca PyAutoGui, para exibir a posição do cursor do mouse;
* **copy** da biblioteca PyPerClip, para copiar um texto;
* **click** da biblioteca PyAutoGui, para simular o clique do mouse;


Primeiro, vamos realizar o download das bibliotecas. Como o ambiente de acesso está sendo na nuvem, através do Google Colab, devemos iniciar essa linha de comando toda vez que executarmos o código, pois ao término da execução e fechamento da aba, todo o progresso é reiniciado. Se caso contrário, esse código estiver sendo executado no Jupyter Notebook, a célula de download da bibliteca pode ser executada apenas uma única vez.

## Download e Importação das bibliotecas

In [1]:
# deve ser executada apenas uma unica vez para o ambiente baixar as bibliotecas para o kernel
# download das bibliotecas do projeto
# !pip install pyperclip
# !pip install pyautogui

In [2]:
# importando as bibliotecas para a execução do codigo
import pyautogui
import pyperclip
import time
import pandas as pd

## Automatização com Python

Como cada computador possui um único tamanho de tela, vamos verificar qual o tamanho está sendo executado nesta execução do código.

### Tamanho do Monitor

In [3]:
# obtendo o tamanho da tela
largura, altura = pyautogui.size()
print("Monitor: {} x {}".format(largura, altura))

Monitor: 1366 x 768


A máquina onde está sendo feito e executado o código está em um computador de tela com tamanho $1366 \times 768.$ Isso será crucial pois a posição do mouse será realizada em cima dos parâmetros de altura e largura da tela.

### Acessando site e realizando download de um arquivo

Agora, vamos realizar a etapa inicial do projeto de automatização. Nosso intuito é abrir uma aba do navegador, colar um endereço para uma determinada página e assim seguir algumas ações. A página que estamos interessados em acessar é o **Gmail,** para escrever e enviar um email para um destinatário específico.

**Obs.: Essa versão só será possível se o email já estiver logado automaticamente na máquina.** 

Primeiro, vamos fazer com que o código abra a aba do navegador e cole um link para navegação. Depois vamos navegar até a página do **Drive**, onde o arquivo xlsx está, e realizar o download.

Ao utilizar a função *click*, é necessário passar a posição $x$ e $y$ do mouse. Mas, devido aos diferentes tamanhos de tela nos computadores, foi feita uma relação da posição real com o tamanho da tela do computador.

In [4]:
pyautogui.PAUSE =  1 # definindo uma pausa de 2 seg para cada comando do pyautogui

# abrindo uma nova janela anônima, escrevendo endereço e acessando
#pyautogui.hotkey("ctrl", "shift", "n") -> abrir janela anonima mas o gmail não vai estar logado
pyautogui.hotkey("ctrl", "t")
pyautogui.click(x = largura * 0.354, y = altura * 0.0625)
pyautogui.write("https://docs.google.com/spreadsheets/d/1KPeP7RmgZfYoB8pUJ1JipAYq0U8X5sPS/edit?usp=sharing&ouid=100677716376606374765&rtpof=true&sd=true", interval=0.03)
pyautogui.press("enter") 
time.sleep(10) # adicionando 10 seg parados para garantir o carregamento da pag

# download de arquivo csv
pyautogui.click(x = 0.0688 * largura, y = 0.151 * altura)
pyautogui.moveTo(x = 0.147877 * largura, y = 0.513 * altura)
pyautogui.click(x = 0.3887 * largura, y = 0.5195 * altura)
time.sleep(5)

Os dados mostrados abaixos são fictícios, representando a venda de produtos em lojas  em lojas localizadas em shoppings. O nosso objetivo é realizar o cálculo da quantidade de produtos vendidos e faturamento anual das vendas dos produtos de todas as lojas. Para isso, é necessário somar os valores das colunas `Quantidade` e `Valor Final`, representado pelo produto da quantidade e o valor unitário do produto.

In [5]:
# realizando análise do faturamento mensal
dados = pd.read_excel("/home/particular/Downloads/Vendas - Dez.xlsx")
display(dados)
faturamento = dados['Valor Final'].sum()
quantidade = dados['Quantidade'].sum()

Código Venda       Data                     ID Loja           Produto  \
0            65014 2019-12-01            Shopping Morumbi    Sunga Listrado   
1            65014 2019-12-01            Shopping Morumbi   Casaco Listrado   
2            65016 2019-12-01           Iguatemi Campinas   Sapato Listrado   
3            65016 2019-12-01           Iguatemi Campinas            Casaco   
4            65017 2019-12-01          Shopping SP Market        Gorro Liso   
...            ...        ...                         ...               ...   
7084         69996 2019-12-26  Center Shopping Uberlândia    Short Listrado   
7085         69996 2019-12-26  Center Shopping Uberlândia           Mochila   
7086         69996 2019-12-26  Center Shopping Uberlândia  Pulseira Estampa   
7087         69997 2019-12-26           Ribeirão Shopping   Camisa Listrado   
7088         69997 2019-12-26           Ribeirão Shopping       Short Linho   

      Quantidade  Valor Unitário  Valor Final  
0              5             114          570  
1              1             269          269  
2              2             363          726  
3              1             250          250  
4              3              92          276  
...          ...             ...          ...  
7084           2             102          204  
7085           4             270         1080  
7086           1              87           87  
7087           1             108          108  
7088           2             133          266  

[7089 rows x 7 columns]

### Gmail - Acessando, escrevendo mensagem e enviando 

Nessa etapa final, vamos automatizar o envio de um email, com as informações obtidas do conjunta de dados a fim de informar, por exemplo, um gestor da área ou gerente da empresa. Vamos acessar o gmail, escrever um email e enviar, para um destinatário definido e assunto. Isso tudo sem encostar no teclado, totalmente automatizado com python.

A execução do código está sendo em uma máquina com sistema operacional **Linux.** Ao utilizar a função copy, da biblioteca *PyPerClip*, o mesmo apresentava o seguinte erro: **"QXcbClipboard::setMimeData: Cannot set X11 selection owner"**. Foi observado que no so Linux não era possível copiar um texto para a área de transferência. A execução apresentava que não era possível definir o modo de área de transferência do proprietário da seleção, recurso de so para armazenamento de pequenas quantidades de dados para transferência entre documentos ou aplicativo. Com isso, tivemos que adotar uma outra biblioteca para contornar esses problemas. 

A primeira tentativa seria realizar a escrita através do comando write, da biblioteca pyautogui, mas apresentou erros de escrita em caracteres especiais. Com isso, foi adotado o uso da classe [Keyboard](https://pynput.readthedocs.io/en/latest/keyboard.html), da biblioteca [pynput](https://pynput.readthedocs.io/en/latest/index.html). Essa classe assume o controle total do seu teclado da máquina, com a possibilidade de simular a escrita, sem a alteração com os caracteres especiais do teclado.

In [6]:
#!pip install pynput
from pynput.keyboard import Controller

In [7]:
# Abrindo uma nova aba e acessando o gmail
pyautogui.hotkey('ctrl', 't')
pyautogui.click(x = largura * 0.354, y = 768 * 0.0625)
pyautogui.write("https://mail.google.com/mail/u/0/#inbox", interval=0.08)
pyautogui.press("enter")

# esperando a pagina carregar e clicando no botao escrever
time.sleep(10)
pyautogui.click(x = largura * 0.05, y = altura * 0.225)

# preenchendo destinatario, assunto 
pyautogui.write('gerente-teste@outlook.com')
pyautogui.press('tab', presses=2)
time.sleep(1)
keyboard = Controller()
keyboard.type('Informação Anual da Companhia') #pyautogui.copy('Informação Anual da Companhia', interval=0.08) -> pyautogui.hotkey('ctrl', 'v')
pyautogui.press('tab')
time.sleep(2)

# escrevendo mensagem do corpo do email
texto = f'''
    Prezadxs, bom dia!
    
    Nosso ano foi um espetáculo. Vendemos um total de {quantidade} produtos em nosso estoque, separados nas diversas lojas localizadas nos shoppings da região.
    O faturamento total anual foi de R$ {faturamento:,.2f}.
    
    Atenciosamente,
    
    Guirra Datatec.
'''

# escrevendo automaticamente
keyboard.type(texto) # ou pyperclip.copy(texto) -> pyautogui.hotkey('ctrl', 'v')

# enviando email
pyautogui.hotkey('ctrl', 'enter')